Transient Pool

ryanolson opened the pull request:
```cpp
/**
 * @brief ReusablePool of memory::buffers that are used as reusable reference-counted monotonic memory resources
 *
 * TransientPool is a ReusablePool of memory::buffers from which smaller buffers are allocated similar to a monotonic
 * memory resource, i.e. pointer pusing stack; however the TransientBuffer or Transient<T> object pulled from the pool
 * hold a SharedReusable<memory::buffer> which keeps the entire monotonic stack from returning to the ReusablePool
 * until all transient objects created on a given stack are deallocated.
 *
 * Allocation of Transient object should be incredibly fast; even faster than the Reusable/SharedResuable on which they
 * are based, since a single Reusable<memory::buffer> might back 10s-1000s of allocations depending on size.
 *
 * It is critical that all Transient object allocated from a pool have similar life cycles
 */
```

dagardner-nv commented on src/internal/memory/transient_pool.hpp:
```diff
@@ -0,0 +1,148 @@
+    DEFAULT_MOVEABILITY(TransientBuffer);
+
+    void* data();
+    std::size_t bytes() const;
```
Maybe just a comment of what bytes represents here, is it the size of `data` or is it the total size of `TransientBuffer`>

Also is `addr` null terminated?

dagardner-nv commented on src/internal/memory/transient_pool.hpp:
```diff
@@ -0,0 +1,148 @@
+ * @brief ReusablePool of memory::buffers that are used as reusable reference-counted monotonic memory resources
+ *
+ * TransientPool is a ReusablePool of memory::buffers from which smaller buffers are allocated similar to a monotonic
+ * memory resource, i.e. pointer pusing stack; however the TransientBuffer or Transient<T> object pull from the pool
```
```suggestion
 * memory resource, i.e. pointer pushing stack; however the TransientBuffer or Transient<T> object pull from the pool
```

dagardner-nv reviewed (changes requested):
Looks good, but can you add a few more comments to the header?

ryanolson commented on src/internal/memory/transient_pool.hpp:
```diff
@@ -0,0 +1,148 @@
+  public:
+    Transient(TransientBuffer&& buffer) : TransientBuffer(std::move(buffer))
+    {
+        CHECK_LT(sizeof(T), bytes());
```
the buffer should be allocated to be a bit larger than sizeof(T) to allow for proper alignment.

it should be

CHECK_LE(sizeof(T) + alignof(T), bytes());

ryanolson commented on src/internal/memory/transient_pool.hpp:
```diff
@@ -0,0 +1,148 @@
+  public:
+    Transient(TransientBuffer&& buffer) : TransientBuffer(std::move(buffer))
+    {
+        CHECK_LT(sizeof(T), bytes());
```
```suggestion
        CHECK_LE(sizeof(T) + alignof(T), bytes());
```

ryanolson commented:
@dagardner-nv - i added explicit move constructors and assignment operators since the default ones do not play nice with raw pointers.

I also added the ability to pass arguments for the `await_object<T>(...)` call.

Let me know if you think of any thing I might have missed.
